src/python-common/ceph/deployment/service_spec.py

```diff
@@ -408,13 +408,15 @@ def __init__(self,
                  placement: Optional[PlacementSpec] = None,
                  count: Optional[int] = None,
                  unmanaged: bool = False,
+                 preview_only: bool = False,
```

Comment:
> Like for example, as if ceph orch apply -i rgw.yaml --dry-run would actually overwrite the existing spec.

That's true.. So it's should probably be saved _alongside_ the existing spec?

> Hm. this feels very wrong to me, as if we're mixing things here that should not be mixed.

I couldn't find another way to to prevent the specs from deploying besides introducing a new attribute. Any idea how to make it work?

Reply:
> > Like for example, as if ceph orch apply -i rgw.yaml --dry-run would actually overwrite the existing spec.
> 
> That's true.. So it's should probably be saved _alongside_ the existing spec?

our data structure (thankfully) mandates that we have a **single** spec per service. I think just storing a boolean alongside the spec is not enough to prevent `--dry-run` from overwriting existing specs. I.e.

```bash
# apply the first version:
$ echo <<EOF | ceph orch apply -i
service_type: rgw
service_name: realm.zone
placment:
  hosts:
    - host1
    - host2
# generate a preview:
$ echo <<EOF | ceph orch apply --dry-run -i -
service_type: rgw
service_name: realm.zone
placment:
  label: myzone
# validate that the spec was not changed:
$ ceph orch ls --export --service_name rgw.realm.zone
service_type: rgw
service_name: realm.zone
placment:
  hosts:
    - host1
    - host2
```

> 
> > Hm. this feels very wrong to me, as if we're mixing things here that should not be mixed.
> 
> I couldn't find another way to to prevent the specs from deploying besides introducing a new attribute. Any idea how to make it work?

I think we need to split active (managed and unamanged) services and preview specs.

Reply:
right, so let's agree to save *one* spec that is specifically for previewing that doesn't overwrite the 'acting' spec. omitting the `--dry-run` flag swaps the `preview-spec` with the `acting-spec`.

sounds like a plan :+1: